src/server/oauth2/OAuth2Client.ts

```diff
@@ -89,18 +140,71 @@ export class OAuth2Client {
         http: IHttp,
         persis: IPersistence,
     ): Promise<IApiResponse> {
-        persis.createWithAssociations({
-            token: 'asdkajsdhkjasd',
-            expiresAt: 19001239120,
-            refreshToken: 'sdjadhakjdhkahdkajshd',
-        }, [
-            // How do we find user ID? Via `state` query param?
-            new RocketChatAssociationRecord(RocketChatAssociationModel.USER, request.query.state),
-            new RocketChatAssociationRecord(RocketChatAssociationModel.MISC, `${this.config.alias}-oauth-connection`),
-        ]);
+        try {
+            const siteUrl = await this.app
+                .getAccessors()
+                .environmentReader.getServerSettings()
+                .getValueById('Site_Url');
+            const accessTokenUrl = this.config.accessTokenUri;
+            const redirectUri = this.app
+                .getAccessors()
+                .providedApiEndpoints[0].computedPath.substring(1);
+
+            const {
+                config: { clientId, clientSecret },
+            } = this;
+
+            const {
+                query: { code, state },
+            } = request;
+
+            const url = new URL(accessTokenUrl, siteUrl);
+
+            url.searchParams.set('client_id', clientId);
+            url.searchParams.set('redirect_uri', `${siteUrl}${redirectUri}`);
+            url.searchParams.set('code', code);
+            url.searchParams.set('client_secret', clientSecret);
+            url.searchParams.set('grant_type', IGrantType.AuthorizationCode);
 
-        return {
-            status: 200,
-        };
+            const { content, statusCode } = await http.post(url.href, {
+                headers: { Accept: 'application/json' },
+            });
+
+            const { access_token, expires_in, refresh_token } = JSON.parse(
+                content as string,
+            );
+
+            persis.createWithAssociations(
+                {
+                    token: access_token,
+                    expiresAt: expires_in || '',
+                    refreshToken: refresh_token || '',
+                },
+                [
+                    new RocketChatAssociationRecord(
+                        RocketChatAssociationModel.USER,
+                        state,
+                    ),
+                    new RocketChatAssociationRecord(
+                        RocketChatAssociationModel.MISC,
+                        `${this.config.alias}-oauth-connection`,
+                    ),
+                ],
+            );
+
+            return {
```

Comment:
I'm wondering if we should have a callback here for the App to do something when the user authorizes the connection. Something along the lines of 

```
autorizationCallback: (token: IOAuthToken, user: IUser /*, accessors */): Promise<{
    responseContent?: string; // Content to be shown to the user maybe
} | undefined>
```

This function would have to be sent as a parameter to the `setup` method and stored as a property of the `config` object.

What do you think? @AllanPazRibeiro @thassiov